core/src/main/scala/kafka/log/UnifiedLog.scala

```
@@ -600,6 +600,11 @@ class UnifiedLog(@volatile var logStartOffset: Long,
    */
   def hasOngoingTransaction(producerId: Long, producerEpoch: Short): Boolean = lock synchronized {
     val entry = producerStateManager.activeProducers.get(producerId)
+    // With transactions V2, if we see a future epoch, we are likely in the process of completing the previous transaction.
+    // Return early with ConcurrentTransactionsException until the transaction completes.
+    if (entry != null && entry.currentTxnFirstOffset.isPresent && entry.producerEpoch() < producerEpoch)
```

It reminds me of an optimation https://issues.apache.org/jira/browse/KAFKA-18042
Basically `entry.producerEpoch() > producerEpoch` is also illegal and can be rejected earlier instead of sending the add partition verification. I wonder if we can also throw an exception(ProducerFenced) here?

---

I would like to try to keep this change simple due to the code freeze. But yes, we could do that in a followup. 👍